File: Sources/Nexus/Support/AppState.swift
import SwiftUI
import NexusCore
import UserNotifications
import ServiceManagement

enum SidebarItem: String, CaseIterable, Identifiable, Hashable {
    case today, review, files, projects, rules, tasks, insights, activity, connectors, access, developer
    var id: String { rawValue }
    var title: String {
        switch self {
        case .today: return "Today"
        case .review: return "Review Queue"
        case .files: return "Files"
        case .projects: return "Projects"
        case .rules: return "Rules & Automations"
        case .tasks: return "Tasks & Schedule"
        case .insights: return "Insights"
        case .activity: return "Activity"
        case .connectors: return "Connectors"
        case .access: return "System Access"
        case .developer: return "Developer"
        }
    }
    var symbol: String {
        switch self {
        case .today: return "sun.horizon"
        case .review: return "tray.full"
        case .files: return "doc.text.magnifyingglass"
        case .projects: return "square.stack.3d.up"
        case .rules: return "point.3.connected.trianglepath.dotted"
        case .tasks: return "calendar.day.timeline.left"
        case .insights: return "chart.xyaxis.line"
        case .activity: return "clock.arrow.circlepath"
        case .connectors: return "puzzlepiece.extension"
        case .access: return "lock.shield"
        case .developer: return "chevron.left.forwardslash.chevron.right"
        }
    }
    static func from(_ name: String) -> SidebarItem? {
        let n = name.lowercased()
        if n.contains("review") { return .review }
        if n.contains("insight") { return .insights }
        if n.contains("rule") { return .rules }
        if n.contains("project") { return .projects }
        if n.contains("task") || n.contains("schedule") { return .tasks }
        if n.contains("activity") { return .activity }
        if n.contains("connector") { return .connectors }
        if n.contains("file") { return .files }
        if n.contains("today") { return .today }
        if n.contains("access") || n.contains("permission") { return .access }
        return nil
    }
}

/// Main-actor bridge between the engine (background) and SwiftUI.
@MainActor
final class AppState: ObservableObject {
    static let shared = AppState()

    let engine: NexusEngine
    let api: APIServer
    let remote: RemoteServer
    @Published var remoteDevices: [RemoteServer.Device] = []
    @Published var remoteRunning = false

    @Published var selection: SidebarItem? = .today
    @Published var status: EngineStatus = .idle
    @Published var settings: NexusSettings
    @Published var reviewItems: [ReviewItem] = []
    @Published var insights: [Insight] = []
    @Published var rules: [Rule] = []
    @Published var conflicts: [RuleConflict] = []
    @Published var projects: [Project] = []
    @Published var jobs: [Job] = []
    @Published var schedules: [Schedule] = []
    @Published var events: [ActivityEvent] = []
    @Published var fileCount = 0
    @Published var focus: FocusSession?
    @Published var paused = false
    @Published var llmName = "Checking…"
    @Published var recentCommands: [String] = []
    @Published var showOnboarding = false
    @Published var ruleDraft: Rule?          // opened in the rules editor (e.g. from an insight)
    @Published var selectedProjectId: String?
    @Published var toast: String?
    @Published var showShortcuts = false

    private var pending = Set<String>()
    private var flushScheduled = false
    private var statusWork: DispatchWorkItem?
    private var workingSince: Date?
    private var pendingStatus: EngineStatus?

    private init() {
        let store: NexusStore
        do { store = try NexusStore() } catch {
            fatalError("Nexus could not open its database: \(error)")
        }
        engine = NexusEngine(store: store)
        api = APIServer(engine: engine)
        remote = RemoteServer(api: api, store: store)
        api.remote = remote
        settings = engine.settings
    }

    func start() {
        engine.notifier = { title, body, important in
            Task { @MainActor in AppState.shared.deliverNotification(title: title, body: body, important: important) }
        }
        engine.bus.subscribe { event in
            switch event {
            case .storeChanged(let entity): Task { @MainActor in AppState.shared.invalidate(entity) }
            case .status(let s): Task { @MainActor in AppState.shared.setStatus(s) }
            case .focusChanged: Task { @MainActor in AppState.shared.invalidate("focus") }
            default: break
            }
        }
        // Snapshot taken when Nexus is summoned (before it activates); safe to read from any thread.
        engine.contextSelection = { ContextCapture.selection }
        engine.start()
        if settings.apiEnabled { api.start(port: settings.apiPort) }
        remote.onChange = { Task { @MainActor in AppState.shared.remoteDevices = AppState.shared.remote.devices; AppState.shared.remoteRunning = AppState.shared.remote.isRunning } }
        if settings.remoteEnabled { remote.start() }
        remoteDevices = remote.devices
        requestNotificationPermission()
        reloadAll()
        showOnboarding = !settings.onboardingComplete
        Task { llmName = await engine.llm.providerName() }
    }

    /// Status is debounced so bursts of tiny background work don't make the menu bar flicker:
    /// "working" only shows after 0.8s of continuous activity and stays at least 1.5s.
    func setStatus(_ s: EngineStatus) {
        if s == pendingStatus { return }          // already scheduled — don't restart the debounce
        if s == status { statusWork?.cancel(); pendingStatus = nil; return }
        pendingStatus = s
        statusWork?.cancel()
        let apply = DispatchWorkItem { [weak self] in
            guard let self else { return }
            self.pendingStatus = nil
            guard self.status != s else { return }
            self.workingSince = s == .working ? Date() : nil
            self.status = s
        }
        statusWork = apply
        let delay: TimeInterval
        if s == .working { delay = 0.8 }
        else if let since = workingSince { delay = max(0, 1.5 - Date().timeIntervalSince(since)) }
        else { delay = 0 }
        DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: apply)
    }

    func invalidate(_ entity: String) {
        pending.insert(entity)
        guard !flushScheduled else { return }
        flushScheduled = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { [weak self] in
            guard let self else { return }
            self.flushScheduled = false
            let entities = self.pending
            self.pending.removeAll()
            self.reload(entities)
        }
    }

    func reloadAll() { reload(["files", "rules", "projects", "jobs", "schedules", "events", "insights", "review", "settings", "focus", "commands"]) }

    private func reload(_ e: Set<String>) {
        let store = engine.store
        if e.contains("review") || e.contains("files") { reviewItems = store.reviewItems() }
        if e.contains("insights") { insights = store.insights() }
        if e.contains("rules") { rules = store.rules(); conflicts = engine.ruleEngine.analyzeConflicts(rules) }
        if e.contains("projects") { projects = store.projects() }
        if e.contains("jobs") { jobs = store.jobs(limit: 300) }
        if e.contains("schedules") { schedules = store.schedules() }
        if e.contains("events") { events = store.events(limit: 400); recentCommands = store.commandHistory() }
        if e.contains("files") { fileCount = store.fileCount() }
        if e.contains("settings") {
            settings = engine.settings
            // Apply changes made outside the Settings UI (API, CLI, iPhone)
            if settings.remoteEnabled != remote.isRunning { settings.remoteEnabled ? remote.start() : remote.stop() }
            if settings.apiEnabled && api.port == 0 { api.start(port: settings.apiPort) }
        }
        focus = engine.focus
        paused = engine.paused
        if engine.status != status { setStatus(engine.status) }
        WidgetBridge.publish(self)
    }

    // MARK: Actions

    func saveSettings(_ s: NexusSettings) {
        let apiChanged = s.apiEnabled != settings.apiEnabled || s.apiPort != settings.apiPort
        let hotkeyChanged = s.paletteHotkey != settings.paletteHotkey || s.voiceHotkey != settings.voiceHotkey
        engine.updateSettings(s)
        settings = s
        if apiChanged { s.apiEnabled ? api.start(port: s.apiPort) : api.stop() }
        if s.remoteEnabled != remote.isRunning { s.remoteEnabled ? remote.start() : remote.stop(); remoteRunning = remote.isRunning }
        if hotkeyChanged { PaletteController.shared.registerHotkeys(palette: s.paletteHotkey, voice: s.voiceHotkey) }
        NSApp.setActivationPolicy(s.showDockIcon ? .regular : .accessory)
        HotbarController.shared.apply(s.hotbar)
        Task { llmName = await engine.llm.providerName() }
    }

    func runCommand(_ text: String) {
        Task {
            let plan = await engine.plan(text)
            let result = await engine.execute(plan)
            showToast(result.message)
            if let nav = result.navigate.flatMap(SidebarItem.from) { selection = nav; openMainWindow() }
        }
    }

    /// Onboarding is shown exactly once — even if the sheet is closed without finishing.
    func markOnboardingSeen() {
        guard !settings.onboardingComplete else { return }
        var s = settings
        s.onboardingComplete = true
        saveSettings(s)
    }

    func openSettings() {
        NSApp.activate(ignoringOtherApps: true)
        if #available(macOS 14.0, *) { NSApp.sendAction(Selector(("showSettingsWindow:")), to: nil, from: nil) }
        else { NSApp.sendAction(Selector(("showPreferencesWindow:")), to: nil, from: nil) }
    }

    func setPaused(_ p: Bool) { engine.setPaused(p); paused = p }

    func showToast(_ message: String) {
        toast = message
        DispatchQueue.main.asyncAfter(deadline: .now() + 4) { [weak self] in if self?.toast == message { self?.toast = nil } }
    }

    func openMainWindow(_ item: SidebarItem? = nil) {
        if let item { selection = item }
        NSApp.activate(ignoringOtherApps: true)
        if let w = NSApp.windows.first(where: { $0.identifier?.rawValue.hasPrefix("main") == true || $0.title == "Nexus" }) {
            w.makeKeyAndOrderFront(nil)
        } else {
            NotificationCenter.default.post(name: .openMainWindow, object: nil)
        }
    }

    // MARK: Notifications

    private var notificationsAvailable: Bool { Bundle.main.bundleIdentifier != nil && Bundle.main.bundlePath.hasSuffix(".app") }

    func requestNotificationPermission() {
        guard notificationsAvailable else { return }
        let center = UNUserNotificationCenter.current()
        center.requestAuthorization(options: [.alert, .sound, .badge]) { _, _ in }
        let review = UNNotificationAction(identifier: "review", title: "Open Review Queue", options: [.foreground])
        let insights = UNNotificationAction(identifier: "insights", title: "Show Insights", options: [.foreground])
        center.setNotificationCategories([UNNotificationCategory(identifier: "nexus", actions: [review, insights], intentIdentifiers: [])])
    }

    func deliverNotification(title: String, body: String, important: Bool) {
        guard notificationsAvailable else { showToast("\(title): \(body)"); return }
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.categoryIdentifier = "nexus"
        if important { content.sound = .default; content.interruptionLevel = .timeSensitive } else { content.interruptionLevel = .passive }
        UNUserNotificationCenter.current().add(UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil))
    }

    // MARK: Launch at login

    var launchAtLogin: Bool {
        get { SMAppService.mainApp.status == .enabled }
        set {
            do { newValue ? try SMAppService.mainApp.register() : try SMAppService.mainApp.unregister() }
            catch { showToast("Launch at login requires running the bundled Nexus.app (\(error.localizedDescription))") }
            objectWillChange.send()
        }
    }
}

extension Notification.Name {
    static let openMainWindow = Notification.Name("NexusOpenMainWindow")
}

File: windows/src/Nexus.App/Support/AppState.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Threading;
using Nexus.Core;

namespace Nexus.App;

public class Observable : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;
    protected bool Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        return true;
    }
    protected void Raise([CallerMemberName] string? name = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

public record ReviewRow(ReviewItem Item)
{
    public string Name => System.IO.Path.GetFileName(Item.Path);
    public string From => Paths.Abbreviate(Paths.FolderOf(Item.Path));
    public string Destination => Item.SuggestedDestination is { } d ? Paths.Abbreviate(d) : "—";
    public string Confidence => $"{(int)(Item.Confidence * 100)}%";
    public string Reason => string.Join(" · ", Item.Reasons);
    public string Tags => Item.SuggestedTags.Count == 0 ? "" : string.Join("  ", Item.SuggestedTags.Select(t => "#" + t));
    public List<string> Alternatives => Item.Alternatives.Select(Paths.Abbreviate).ToList();
}

public record InsightRow(Insight Item)
{
    public string Title => Item.Title;
    public string Detail => Item.Detail;
    public string Severity => Item.Severity.ToString().ToUpperInvariant();
    public bool HasFix => !string.IsNullOrEmpty(Item.Command);
    public string FixLabel => Item.Kind switch { InsightKind.habit => "Create rule", InsightKind.conflicts => "Review rules", InsightKind.deadlineSoon => "Start focus", _ => "Fix it" };
}

public class RuleRow(Rule rule) : Observable
{
    public Rule Rule { get; } = rule;
    public string Name => Rule.Name;
    public string Summary => Rule.Summary;
    public string Hits => Rule.HitCount == 0 ? "never fired" : $"{Rule.HitCount} hits";
    public bool Enabled
    {
        get => Rule.Enabled;
        set { Rule.Enabled = value; AppState.Shared.Engine.Store.SaveRule(Rule); AppState.Shared.Engine.RestartWatcher(); Raise(); }
    }
}

public record EventRow(ActivityEvent Item)
{
    public string Time => Item.Timestamp.ToString(Item.Timestamp.Date == DateTime.Today ? "HH:mm" : "MMM d HH:mm");
    public string Message => Item.Message;
    public string Kind => Item.Kind.ToString();
    public bool Undoable => Item.Undo != null && !Item.Undone;
    public double Opacity => Item.Undone ? 0.45 : 1;
}

public record JobRow(Job Item)
{
    public string Name => Item.Name;
    public string Status => Item.Status.ToString();
    public string Detail => Item.Error ?? Item.ResultSummary ?? (Item.Status == JobStatus.scheduled ? $"at {Item.ScheduledFor:MMM d HH:mm}" : "");
    public string When => Item.CreatedAt.ToString("MMM d HH:mm");
    public bool CanRetry => Item.Status == JobStatus.failed;
}

public record ProjectRow(Project Item, int Files, long Bytes)
{
    public string Name => Item.Name;
    public string Stats => $"{Files} files · {Text.FormatBytes(Bytes)}" + (Item.Deadline is { } d ? $" · due {d:MMM d}" : "");
    public string Keywords => string.Join(", ", Item.Keywords);
    public string Color => Item.Color;
}

public record FileRow(FileRecord Item)
{
    public string Name => Item.Name;
    public string Folder => Paths.Abbreviate(Item.Folder);
    public string Meta => string.Join(" · ", new[] { Item.DocType, Item.Topics.Count > 0 ? string.Join(", ", Item.Topics.Take(3)) : null, Item.Tags.Count > 0 ? string.Join(" ", Item.Tags.Select(t => "#" + t)) : null }.Where(x => !string.IsNullOrEmpty(x)));
    public string Snippet => Item.Summary ?? Item.Snippet;
}

public record DeviceRow(RemoteServer.Device Item)
{
    public string Name => Item.Name;
    public string Seen => Item.LastSeen is { } s ? "last seen " + Nexus.Core.Text.Relative(s) : "paired " + Item.PairedAt.ToString("MMM d");
}

/// Main-thread bridge between the engine (background) and WPF.
public class AppState : Observable
{
    public static AppState Shared { get; } = new();

    public NexusEngine Engine { get; }
    public ApiServer Api { get; }
    public RemoteServer Remote { get; }

    public ObservableCollection<ReviewRow> Reviews { get; } = [];
    public ObservableCollection<InsightRow> Insights { get; } = [];
    public ObservableCollection<RuleRow> Rules { get; } = [];
    public ObservableCollection<EventRow> Events { get; } = [];
    public ObservableCollection<JobRow> Jobs { get; } = [];
    public ObservableCollection<string> Upcoming { get; } = [];
    public ObservableCollection<ProjectRow> Projects { get; } = [];
    public ObservableCollection<DeviceRow> Devices { get; } = [];
    public ObservableCollection<string> Conflicts { get; } = [];
    public ObservableCollection<string> RecentCommands { get; } = [];

    string status = "Idle", llm = "Checking…", toast = "", focusText = "", ticker = "Ready";
    int files, review, rulesCount, insightCount, filedToday;
    bool paused, remoteRunning;
    EngineStatus engineStatus;

    public string StatusText { get => status; set => Set(ref status, value); }
    public EngineStatus EngineStatus { get => engineStatus; set => Set(ref engineStatus, value); }
    public string LlmName { get => llm; set => Set(ref llm, value); }
    public string Toast { get => toast; set => Set(ref toast, value); }
    public string FocusText { get => focusText; set => Set(ref focusText, value); }
    public string Ticker { get => ticker; set => Set(ref ticker, value); }
    public int FileCount { get => files; set => Set(ref files, value); }
    public int ReviewCount { get => review; set => Set(ref review, value); }
    public int RulesCount { get => rulesCount; set => Set(ref rulesCount, value); }
    public int InsightCount { get => insightCount; set => Set(ref insightCount, value); }
    public int FiledToday { get => filedToday; set => Set(ref filedToday, value); }
    public bool Paused { get => paused; set => Set(ref paused, value); }
    public bool RemoteRunning { get => remoteRunning; set => Set(ref remoteRunning, value); }
    public NexusSettings Settings => Engine.Settings;

    public event Action<string>? Navigate;
    readonly HashSet<string> pending = [];
    readonly object pendingLock = new();
    DispatcherTimer? flush;
    DispatcherTimer? statusTimer;
    EngineStatus? pendingStatus;

    AppState()
    {
        var store = new NexusStore();
        Engine = new NexusEngine(store);
        Api = new ApiServer(Engine);
        Remote = new RemoteServer(Api, store);
        Api.Remote = Remote;
    }

    public void Start()
    {
        Engine.Notifier = (t, b, important) => Application.Current.Dispatcher.BeginInvoke(() => Tray.Shared.Notify(t, b, important));
        Engine.StoreChanged += e => Invalidate(e);
        Engine.StatusChanged += s => Application.Current.Dispatcher.BeginInvoke(() => SetStatus(s));
        Engine.ContextSelection = ExplorerSelection.Current;
        Remote.OnChange = () => Application.Current.Dispatcher.BeginInvoke(() => { RemoteRunning = Remote.IsRunning; Reload(["remote"]); });
        Engine.Start();
        if (Settings.ApiEnabled) Api.Start(Settings.ApiPort);
        if (Settings.RemoteEnabled) Remote.Start();
        RemoteRunning = Remote.IsRunning;
        ReloadAll();
        _ = RefreshLlm();
    }

    public async Task RefreshLlm() => LlmName = await Engine.Llm.ProviderName();

    public void Stop()
    {
        Engine.Stop(); Api.Stop(); Remote.Stop();
    }

    /// "Working" only shows after 0.8 s of continuous activity and then stays at least 1.5 s — no flicker.
    void SetStatus(EngineStatus s)
    {
        if (pendingStatus == s) return;
        pendingStatus = s;
        statusTimer?.Stop();
        statusTimer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(s == EngineStatus.working ? 800 : 400) };
        statusTimer.Tick += (_, _) =>
        {
            statusTimer.Stop();
            pendingStatus = null;
            EngineStatus = s;
            StatusText = s switch { EngineStatus.working => "Working", EngineStatus.attention => "Needs review", EngineStatus.paused => "Paused", _ => "Idle" };
            Tray.Shared.SetStatus(s, ReviewCount);
        };
        statusTimer.Start();
    }

    public void Invalidate(string entity)
    {
        lock (pendingLock) pending.Add(entity);
        Application.Current?.Dispatcher.BeginInvoke(() =>
        {
            if (flush != null) return;
            flush = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(350) };
            flush.Tick += (_, _) =>
            {
                flush!.Stop(); flush = null;
                HashSet<string> e; lock (pendingLock) { e = [.. pending]; pending.Clear(); }
                Reload(e);
            };
            flush.Start();
        });
    }

    public void ReloadAll() => Reload(["files", "rules", "projects", "jobs", "schedules", "events", "insights", "review", "settings", "focus", "remote"]);

    void Reload(IEnumerable<string> entitiesE)
    {
        var e = entitiesE.ToHashSet();
        var store = Engine.Store;
        if (e.Contains("review") || e.Contains("files")) Replace(Reviews, store.ReviewItems().Select(r => new ReviewRow(r)));
        if (e.Contains("insights")) Replace(Insights, store.Insights().Select(i => new InsightRow(i)));
        if (e.Contains("rules"))
        {
            var rules = store.Rules();
            Replace(Rules, rules.Select(r => new RuleRow(r)));
            Replace(Conflicts, Engine.RuleEngine.AnalyzeConflicts(rules).Select(c => c.Message));
        }
        if (e.Contains("projects") || e.Contains("files")) Replace(Projects, store.Projects(false).Select(p => { var s = store.ProjectStats(p.Id); return new ProjectRow(p, s.count, s.size); }));
        if (e.Contains("jobs") || e.Contains("schedules"))
        {
            Replace(Jobs, store.Jobs(150).Select(j => new JobRow(j)));
            Replace(Upcoming, Engine.Scheduler.Upcoming(20).Select(u => $"{u.at:ddd MMM d · h:mm tt}  —  {u.name}"));
        }
        if (e.Contains("events"))
        {
            var events = store.Events(300);
            Replace(Events, events.Select(x => new EventRow(x)));
            Replace(RecentCommands, store.CommandHistory(8));
            if (events.FirstOrDefault(x => x.Kind is not (EventKind.jobStarted or EventKind.jobCompleted)) is { } last) Ticker = last.Message;
            FiledToday = store.EventCount(EventKind.fileMoved, DateTime.Today);
        }
        if (e.Contains("remote")) Replace(Devices, Remote.Devices.Select(d => new DeviceRow(d)));
        if (e.Contains("settings"))
        {
            Raise(nameof(Settings));
            if (Settings.RemoteEnabled != Remote.IsRunning) { if (Settings.RemoteEnabled) Remote.Start(); else Remote.Stop(); }
            if (Settings.ApiEnabled && Api.Port == 0) Api.Start(Settings.ApiPort);
            Hotbar.Apply(Settings.Hotbar);
        }
        FileCount = store.FileCount();
        ReviewCount = store.ReviewCount();
        RulesCount = store.Rules().Count;
        InsightCount = store.Insights().Count;
        Paused = Engine.Paused;
        FocusText = Engine.Focus is { } f && store.Project(f.ProjectId) is { } fp ? $"Focus: {fp.Name} until {f.EndsAt:h:mm tt}" : "";
        Tray.Shared.SetStatus(EngineStatus, ReviewCount);
        Engine.RefreshStatus();
    }

    static void Replace<T>(ObservableCollection<T> target, IEnumerable<T> items)
    {
        var list = items.ToList();
        target.Clear();
        foreach (var i in list) target.Add(i);
    }

    // MARK: actions

    public void SaveSettings(NexusSettings s)
    {
        var apiChanged = s.ApiEnabled != Settings.ApiEnabled || s.ApiPort != Settings.ApiPort;
        var hotkeys = s.PaletteHotkey != Settings.PaletteHotkey || s.VoiceHotkey != Settings.VoiceHotkey;
        Engine.UpdateSettings(s);
        if (apiChanged) { if (s.ApiEnabled) Api.Start(s.ApiPort); else Api.Stop(); }
        if (s.RemoteEnabled != Remote.IsRunning) { if (s.RemoteEnabled) Remote.Start(); else Remote.Stop(); }
        if (hotkeys) Hotkeys.Register(s);
        LoginItem.Apply(s.LaunchAtLogin);
        ThemeManager.Apply(s.Appearance);
        Hotbar.Apply(s.Hotbar);
        _ = RefreshLlm();
        Raise(nameof(Settings));
    }

    public void MarkOnboardingSeen()
    {
        if (Settings.OnboardingComplete) return;
        var s = Settings; s.OnboardingComplete = true; SaveSettings(s);
    }

    public async Task<CommandResult> RunCommand(string text, bool confirmed = true)
    {
        var plan = await Engine.Plan(text);
        var result = await Engine.Execute(plan);
        ShowToast(result.Message);
        if (result.Navigate != null) Navigate?.Invoke(result.Navigate);
        return result;
    }

    public void GoTo(string page) => Navigate?.Invoke(page);

    DispatcherTimer? toastTimer;
    public void ShowToast(string message)
    {
        Toast = message.Length > 260 ? message[..260] + "…" : message;
        toastTimer?.Stop();
        toastTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(5) };
        toastTimer.Tick += (_, _) => { toastTimer.Stop(); Toast = ""; };
        toastTimer.Start();
    }
}
